### ISSUE TITLE
Distributed memory issue: OSError: [Errno 12] Cannot allocate memory

### ISSUE THREAD

**Issue by rickecon** [PSLmodels/OG-USA#414]
I ran `run_ogusa_example.py` yesterday on my Dell Precision 7820 desktop with the following specs:

* Ubuntu Linux 16.04
* Intel Xeon Gold 6138 2.0GHz, 3.7GHz Turbo, 20C, 10.4GT/s 3UPI, 27M Cache, HT (125W) DDR4-2666 1st
* Intel Xeon Gold 6138 2.0GHz, 3.7GHz Turbo, 20C, 10.4GT/s 3UPI, 27M Cache, HT (125W) DDR4-2666 2nd
* NVIDIA Quadro P1000, 4GB, 4 mDP (7X20T)
* 32GB (4x8GB) 2666MHz DDR4 RDIMM ECC
* M.2 512GB PCIe NVMe Class 40 Solid State Drive
* 2.5" 1TB 7200rpm SATA Hard Drive

So this machine has 40 cores (80 threads), 32GB Memory, an OK GPU, and a big hard drive.

When I ran OG-USA, I got the following error, `OSError: [Errno 12] Cannot allocate memory`. This is during the estimation of the tax functions in the baseline (inside of the `get_micro_data.py`, `get_data()` function). 
```
(ospcdyn) rickecon@evans50core:~/Documents/Economics/OG-USA/run_examples$ python run_ogusa_example.py
You loaded data for 2014.
Tax-Calculator startup automatically extrapolated your data to 2014.
Number of workers =  80
making dir:  ./OUTPUT_BASELINE/SS
making dir:  ./OUTPUT_BASELINE/TPI
In runner, baseline is  True
BW =  10 start year =  2018
BW =  10 begin year =  2018 end year =  2027
Running current law policy baseline
You loaded data for 2014.
Tax-Calculator startup automatically extrapolated your data to 2014.
/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/site-packages/dask/base.py:829: UserWarning: The get= keyword has been deprecated. Please use the scheduler= keyword instead with the name of the desired scheduler like 'threads' or 'processes'
    x.compute(scheduler='threads') 
or with a function that takes the graph and keys
    x.compute(scheduler=my_scheduler_function)
  "The get= keyword has been deprecated. "
year:  2020
year:  2023
year:  2026
year:  2021
year:  2022
year:  2025
year:  2027
year:  2024
Process ForkPoolWorker-9:
Traceback (most recent call last):
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/pool.py", line 108, in worker
    task = get()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/queues.py", line 335, in get
    res = self._reader.recv_bytes()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/connection.py", line 411, in _recv_bytes
    return self._recv(size)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/connection.py", line 386, in _recv
    buf.write(chunk)
MemoryError
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/pool.py", line 405, in _handle_workers
    pool._maintain_pool()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/pool.py", line 246, in _maintain_pool
    self._repopulate_pool()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/pool.py", line 239, in _repopulate_pool
    w.start()
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/context.py", line 277, in _Popen
    return Popen(process_obj)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/home/rickecon/anaconda3/envs/ospcdyn/lib/python3.6/multiprocessing/popen_fork.py", line 66, in _launch
    self.pid = os.fork()
OSError: [Errno 12] Cannot allocate memory
```
The default value for the `num_workers` parameter in the `run_ogusa_example.py` script is 80 on this machine. I tried setting the `num_workers` parameter in  to values of 76, 60, and 16, but I always got this same error.

I searched this error and found the following [StackOverflow thread](https://stackoverflow.com/questions/1367373/python-subprocess-popen-oserror-errno-12-cannot-allocate-memory). The best answer respondent says this error is likely "because of either an honest to God out-of-memory condition... or because `security_vm_enough_memory_mm` failed you while enforcing the overcommit policy. Respondent's suggestions are:

* Switch to a larger instance (i.e., get more memory on my computer).
* Put some coding effort into more effectively controlling your script's memory footprint.

I will eventually do the first option, but we need to immediately think about the second suggestion if the memory is truly being overloaded. It looks like the memory overload is happening in lines 170-174 in `get_micro_data.py`, `get_data()` function. Are there ways that we can pass less data in each instance?

@jdebacker @hdoupe

**Comment by jdebacker** [comment]
@rickecon I haven't seen this before either on my Mac or when running on a node of a Unix cluster (with 20 processors).  You might try setting `num_workers=1` as a test.  If that works, then try 4, 7, 10 (the `run_ogusa_example.py` script can only make use of up to 10 parallel processes, so specifying `num_workers>10` shouldn't result in any change in performance.

**Comment by jdebacker** [comment]
@rickecon Thanks for checking that.  Glad to know you can get at least some parallelization, but it would be ideal to increase.

I'll need to read more about what's going on there.  The data itself is not large (the `puf.csv` file is only about 50mb), so I don't know what is eating up about 7GB+ per process.  I also am surprised your machine maxes out at `num_workers=4` with 32GB of RAM - whereas I haven't  had issue running with the same number of workers on  my laptop with 16GB RAM.

**Comment by hdoupe** [comment]
One thing to note is that the Tax-Calculator objects will be larger than the `puf.csv` file. Another is that the CPS file usually results in twice as much run time and memory usage.

**Comment by jdebacker** [comment]
Closing this issue due to inactivity.  We can reopen if this crops up again on other machines.

**Comment by jdebacker** [comment]
Reopening issue.  Talked with @hdoupe and he's had similar problems running things in parallel in Docker instances on AWS.  He suggested that the Tax Calculator objects are hogging resources. 

One solution to try is to add some garbage collection that Python may be missing.  In particular, all TC objects and the micro data from the TC runs should be deleted when they are no longer needed.

**Comment by rickecon** [comment]
I tested this updated version of `OG-USA` on my 40-core (80-thread) machine. I got the same error with `num_workers = 10` and `num_workers=9`. I'm sure it is just choking on dataset size in memory. The error occurred at the end of the micro data I/O step before the tax function estimation. It ran smoothly and efficiently for `num_workers<=8`.

In the case of `num_workers=8`, the times were the following for the `run_ogusa_example.py` script that involves tax function estimation, steady-state, and time path computation for both the baseline and for the reform. The total run time was ? minutes.  Below is the breakdown.
* Baseline microdata fetch (~2min)
* Baseline tax function estimation (~28min)
* Baseline steady state (<1min)
* Baseline transition path (~?min), each iteration (~2min)
* Reform microdata fetch (~2min)
* Reform tax function estimation (~?min)
* Reform steady state (~?min)
* Reform transition path (~?min), each iteration (~?min)

So basically, better garbage collection and more efficient passing of stripped down datasets buys us 8 parallel threads instead of 4.

@jdebacker @hdoupe

**Comment by hdoupe** [comment]
Very cool! Nice work on the memory management @jdebacker. I'm excited to try this out. Also, what's the RAM on your machine? Did you run it in docker?

**Comment by rickecon** [comment]
@hdoupe I did not run it in Docker. I just ran it in the ogusa-dev conda environment natively. My machine has 32GB RAM (see initial post in this thread).

**Comment by hdoupe** [comment]
Ah, cool. Thanks for the info.

**Comment by rickecon** [comment]
The memory management problem is always critical in advance high performance computing. And this is an area where I am not the most seasoned expert. But it is often not just the amount of RAM available, but also the amount memory on each of the threads.

**Comment by jdebacker** [comment]
[tracemalloc](https://docs.python.org/3/library/tracemalloc.html) may be helpful in finding where we need to focus to reduce these memory issues.

But glad to hear this is running with 8 workers.

**Comment by jdebacker** [comment]
Issue seems to have been fixed.  Closing this thread, but welcome future issues that look at better memory management throughout the model.
